src/shadow.rs

```diff
@@ -0,0 +1,174 @@
+    ///
+    /// An empty field and value 0 mean that there are no password warning
+    /// period.
+    pub warn: libc::c_long,
+    /// Password inactivity period
+    ///
+    /// The number of days after a password has expired (see the maximum
+    /// password age above) during which the password should still be accepted
+    /// (and the user should update their password during the next login).
+    ///
+    /// After expiration of the password and this expiration period is elapsed,
+    /// no login is possible using the current user's password. The user should
+    /// contact their system administrator.
+    pub inactive: libc::c_long,
+    /// Account expiration date
+    ///
+    /// The date of expiration of the account, expressed as the number of days
+    /// since Jan 1, 1970.
+    ///
+    /// Note that an account expiration differs from a password expiration. In
+    /// case of an account expiration, the user shall not be allowed to login.
+    /// In case of a password expiration, the user is not allowed to login using
+    /// their password.
+    ///
+    /// The value 0 should not be used as it is interpreted as either an account
+    /// with no expiration, or as an expiration on Jan 1, 1970.
+    pub expire: libc::c_long,
+}
+
+impl From<&libc::spwd> for Shadow {
+    fn from(spwd: &libc::spwd) -> Shadow {
+        Shadow {
+            name: unsafe { CStr::from_ptr(spwd.sp_namp).to_string_lossy().into_owned() },
+            password: unsafe { CString::new(CStr::from_ptr(spwd.sp_pwdp).to_bytes()).unwrap() },
+            last_change: spwd.sp_lstchg,
+            min: spwd.sp_min,
+            max: spwd.sp_max,
+            warn: spwd.sp_warn,
+            inactive: spwd.sp_inact,
+            expire: spwd.sp_expire,
+        }
+    }
+}
+
+impl Shadow {
+    /// Gets a [`Shadow`] entry for the given username, or returns [`None`].
+    ///
+    /// # Safety
+    /// Care should be taken when this function is used in different threads
+    /// without synchronization. This is because the underlying function used
+    /// ([`getspnam()`][1]) is not thread safe.
+    ///
+    /// [1]: http://man7.org/linux/man-pages/man3/shadow.3.html
+    pub fn from_name(user: &str) -> Option<Shadow> {
+        let c_user = CString::new(user).unwrap();
+
+        let spwd = unsafe { libc::getspnam(c_user.as_ptr()) };
+
+        if spwd.is_null() {
+            None
+        } else {
+            Some(unsafe { Shadow::from(&*spwd) })
+        }
+    }
+
+    /// Returns iterator over all entries in `shadow` file
+    pub fn iter_all() -> ShadowIterator {
+        ShadowIterator::default()
+    }
+}
+
+/// Iterator over `Shadow` entries
+///
+/// # Examples
+/// ```
+/// # use nix::shadow::ShadowIterator;
+/// # use nix::shadow::Shadow;
+/// let shadows = ShadowIterator::default().collect::<Vec<Shadow>>();
+/// println!("There are {} shadow entries", shadows.len());
+/// ```
+///
+/// # Safety
```

Comment:
What about using the reentrant versions `getspent_r` etc to solve the thread-safety problem?

Reply:
I'll do that.

Reply:
I'm currently blocked by https://github.com/rust-lang/libc/pull/1653